### ISSUE TITLE
redundant_closure lint incorrectly emitted when using a `FnMut` as a `FnOnce`

### ISSUE THREAD

**Issue by JoJoJet** [rust-lang/rust-clippy#9369]
### Summary

When passing a `FnMut` closure to a function that expects a `FnOnce`, you must wrap it in a new closure to avoid moving the original. Clippy incorrectly emits a redundant_closure lint when doing this.

### Lint Name

redundant_closure

### Reproducer

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9e821a26232ffb0e3cf5f90511c56c6b

```rust
pub fn ok_or_else<T, E>(mut f: impl FnMut() -> E)
    -> impl FnMut(Option<T>) -> Result<T, E>
{
    move |opt| opt.ok_or_else(|| f())
}
```

I saw this happen:

```text
warning: redundant closure
 --> src/lib.rs:4:31
  |
4 |     move |opt| opt.ok_or_else(|| f())
  |                               ^^^^^^ help: replace the closure with the function itself: `f`
  |
  = note: `#[warn(clippy::redundant_closure)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
```

Following clippy's advice causes:
```text
error[[E0507]](https://doc.rust-lang.org/stable/error-index.html#E0507): cannot move out of `f`, a captured variable in an `FnMut` closure
 --> src/lib.rs:4:31
  |
1 | pub fn ok_or_else<T, E>(mut f: impl FnMut() -> E)
  |                         ----- captured outer variable
...
4 |     move |opt| opt.ok_or_else(f)
  |     ----------                ^ move occurs because `f` has type `impl FnMut() -> E`, which does not implement the `Copy` trait
  |     |
  |     captured by this `FnMut` closure
```

### Version

```text
rustc 1.63.0 (4b91a6ea7 2022-08-08)
binary: rustc
commit-hash: 4b91a6ea7258a947e59c6522cd5898e7c0a6a88f
commit-date: 2022-08-08
```

**Comment by JoJoJet** [comment]
std has `impl<F: FnMut> FnOnce for &mut F`, so it would be better for clippy to suggest

```text
warning: redundant closure
 --> src/lib.rs:4:31
  |
4 |     move |opt| opt.ok_or_else(&mut f)
  |                               ^^^^^^ help: replace the closure with a mutable reference: `&mut f`
  |
  = note: `#[warn(clippy::redundant_closure)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
```
